### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

File: src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs [hunk 1]
```diff
@@ -197,14 +197,18 @@ private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
         {
             if (message.Message.Payload is BlockPayload block)
             {
+                block.Obj.Header.PrecomputeHash(true, true);
```
Comment (zeptin) [inline src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs]: Just an observation - there is potential for a DoS vector here, as the entire block & all transactions are getting hashed before any rules get applied to it.

Perhaps the risk is minor though, as ostensibly we will only request blocks for which we already have a valid header.

@noescape00 what do you think?

Comment (dangershony) [reply src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs]: No the PrecomputeHash method has an option to tell it to lazy load, so only when the hash is calculated we put it in cache.

Comment (zeptin) [reply src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs]: Ah ok, I made an (incorrect) assumption about what `PrecomputeHash` does based on the name.

**Comment by bokobza** [comment]
Did you sync the chains from the beginning?

**Comment by dangershony** [comment]
>Did you sync the chains from the beginning?  

No, I synced STRAT main to 300k and BTC main to 70k.
I will let my nodes run over night with this change

File: src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs]: >as the entire block & all transactions are getting hashed before any rules get applied to it.

only stuff from header gets hashed. Txes affect merkle tree root which is inside of a header.
so should be ok to do it.

**Comment by dangershony** [comment]
I did consider that as well, hmm not a bad idea in the merkle root rule we anyway iterate trx so we could also cache the hash, however that means it wont be cached if called anyplace before hitting that rule, but I don't think this is the case, ok I will make the change.

**Comment by dangershony** [comment]
@noescape00 FYI

**Pull request merged**